File: blender-addon/oh_my_blender/manifest.py
"""Thin Blender extraction layer for Scene Snapshot v2."""

from __future__ import annotations

import json
from pathlib import Path
import unicodedata
from typing import Iterable

import bpy
from mathutils import Quaternion

from .snapshot import (
    UNSUPPORTED_FCURVE_FEATURE,
    UNSUPPORTED_FPS_BASE,
    UNSUPPORTED_LINKED_DATABLOCK,
    assemble_snapshot,
    canonical_quaternion,
    snapshot_revision,
)
from .scene_manifest import (
    build_scene_manifest,
    build_scene_manifest_v3,
    finalize_scene_manifest,
    rational_fps,
)


def _text(value: str) -> str:
    return unicodedata.normalize("NFC", value)


def _vector(values: Iterable[float]) -> list[float]:
    return [float(value) for value in values]


def _object_quaternion(scene_object: bpy.types.Object) -> list[float]:
    if scene_object.rotation_mode == "QUATERNION":
        quaternion = scene_object.rotation_quaternion
    elif scene_object.rotation_mode == "AXIS_ANGLE":
        angle, axis_x, axis_y, axis_z = scene_object.rotation_axis_angle
        quaternion = Quaternion((axis_x, axis_y, axis_z), angle)
    else:
        quaternion = scene_object.rotation_euler.to_quaternion()
    return canonical_quaternion(quaternion)


def _check_linked(scene_object: bpy.types.Object) -> None:
    data = scene_object.data
    if scene_object.library is not None or (data is not None and data.library is not None):
        raise UNSUPPORTED_LINKED_DATABLOCK(
            f"object {scene_object.name!r} uses a linked datablock"
        )


def _object_snapshot(scene_object: bpy.types.Object) -> dict:
    _check_linked(scene_object)
    return {
        "name": _text(scene_object.name),
        "type": _text(scene_object.type),
        "parent": _text(scene_object.parent.name) if scene_object.parent else None,
        "visible": bool(scene_object.visible_get()),
        "location": _vector(scene_object.location),
        "rotationMode": _text(scene_object.rotation_mode),
        "rotationQuaternion": _object_quaternion(scene_object),
        "scale": _vector(scene_object.scale),
    }


def _camera_snapshot(scene_object: bpy.types.Object) -> dict:
    camera = scene_object.data
    return {
        "name": _text(scene_object.name),
        "lens": float(camera.lens),
        "sensorFit": _text(camera.sensor_fit),
        "sensorWidth": float(camera.sensor_width),
        "sensorHeight": float(camera.sensor_height),
        "verticalFovRadians": float(camera.angle_y),
        "clipStart": float(camera.clip_start),
        "clipEnd": float(camera.clip_end),
    }


def _interpolation_values() -> set[str]:
    prop = bpy.types.Keyframe.bl_rna.properties["interpolation"]
    return {item.identifier for item in prop.enum_items}


def animation_fcurves(animation_data: object) -> list[object]:
    """Return only the f-curves associated with an animation data-block's slot."""
    action = animation_data.action
    legacy_fcurves = getattr(action, "fcurves", None)
    if legacy_fcurves is not None and len(legacy_fcurves):
        return list(legacy_fcurves)

    slot = animation_data.action_slot
    result = []
    for layer in action.layers:
        for strip in layer.strips:
            channelbag = strip.channelbag(slot) if hasattr(strip, "channelbag") else None
            if channelbag is not None:
                result.extend(channelbag.fcurves)
                continue
            result.extend(
                fcurve
                for bag in strip.channelbags
                if bag.slot_handle == slot.handle
                for fcurve in bag.fcurves
            )
    return result


def _animation_snapshot(
    object_identity: str,
    target: str,
    animation_data: object,
    *,
    identity_key: str = "objectName",
    include_handle_types: bool = False,
) -> dict | None:
    if animation_data is None:
        return None
    drivers = animation_data.drivers
    if len(drivers):
        raise UNSUPPORTED_FCURVE_FEATURE(
            f"{object_identity!r} {target} animation uses drivers"
        )
    action = animation_data.action
    if action is None:
        return None

    interpolation_values = _interpolation_values()
    easing_defaults = {
        name: bpy.types.Keyframe.bl_rna.properties[name].default
        for name in ("back", "amplitude", "period")
    }
    fcurves = []
    for fcurve in animation_fcurves(animation_data):
        if len(fcurve.modifiers):
            raise UNSUPPORTED_FCURVE_FEATURE(
                f"{object_identity!r} {target} f-curve uses modifiers"
            )
        keyframes = []
        for point in fcurve.keyframe_points:
            if (
                point.easing != "AUTO"
                or point.interpolation not in interpolation_values
                or any(getattr(point, name) != default for name, default in easing_defaults.items())
            ):
                raise UNSUPPORTED_FCURVE_FEATURE(
                    f"{object_identity!r} {target} f-curve uses unsupported easing, "
                    "interpolation, or easing parameters"
                )
            keyframe = {
                "frame": float(point.co.x),
                "value": float(point.co.y),
                "interpolation": point.interpolation,
                "handleLeft": [float(point.handle_left.x), float(point.handle_left.y)],
                "handleRight": [float(point.handle_right.x), float(point.handle_right.y)],
            }
            if include_handle_types:
                keyframe["handleLeftType"] = _text(point.handle_left_type)
                keyframe["handleRightType"] = _text(point.handle_right_type)
            keyframes.append(keyframe)
        fcurves.append(
            {
                "dataPath": _text(fcurve.data_path),
                "arrayIndex": int(fcurve.array_index),
                "keyframes": keyframes,
            }
        )
    return {identity_key: _text(object_identity), "target": _text(target), "fcurves": fcurves}


def _entity_id(entity: object, label: str) -> str:
    entity_id = entity.get("omb.entity_id")
    if not isinstance(entity_id, str):
        raise ValueError(f"{label} is missing omb.entity_id")
    return entity_id


def _manifest_object(scene_object: bpy.types.Object) -> dict:
    snapshot = _object_snapshot(scene_object)
    return {
        "entityId": _entity_id(scene_object, f"object {scene_object.name!r}"),
        "name": snapshot["name"],
        "type": snapshot["type"],
        "parentId": (
            _entity_id(scene_object.parent, f"parent of object {scene_object.name!r}")
            if scene_object.parent
            else None
        ),
        "visible": snapshot["visible"],
        "location": snapshot["location"],
        "rotationQuaternion": snapshot["rotationQuaternion"],
        "scale": snapshot["scale"],
    }


def _manifest_bones(scene_objects: list[bpy.types.Object]) -> list[dict]:
    bones = []
    for scene_object in scene_objects:
        if scene_object.type != "ARMATURE":
            continue
        armature_id = _entity_id(scene_object, f"armature {scene_object.name!r}")
        for bone in scene_object.data.bones:
            location, rotation, scale = bone.matrix_local.decompose()
            bones.append(
                {
                    "entityId": _entity_id(bone, f"bone {bone.name!r}"),
                    "name": _text(bone.name),
                    "armatureObjectId": armature_id,
                    "parentBoneId": (
                        _entity_id(bone.parent, f"parent of bone {bone.name!r}")
                        if bone.parent
                        else None
                    ),
                    "location": _vector(location),
                    "rotationQuaternion": canonical_quaternion(rotation),
                    "scale": _vector(scale),
                }
            )
    return bones


def _stage_manifest_entries(
    scene_objects: list[bpy.types.Object],
    object_ids: dict[bpy.types.Object, str],
) -> tuple[list[dict], list[dict]]:
    primitives = []
    materials = []
    for scene_object in scene_objects:
        object_id = object_ids[scene_object]
        primitive_type = scene_object.get("omb.stage_primitive_type")
        if primitive_type in ("PLANE", "CUBE", "UV_SPHERE"):
            primitives.append({
                "objectId": object_id,
                "primitiveType": primitive_type,
            })
        if scene_object.type != "MESH" or not scene_object.material_slots:
            continue
        material = scene_object.material_slots[0].material
        if (
            material is not None
            and material.get("omb.generated_for_entity_id") == object_id
        ):
            materials.append({
                "objectId": object_id,
                "materialName": _text(material.name),
                "baseColor": _vector(material.diffuse_color),
            })
    return primitives, materials


def _extract_scene_manifest(schema_version: int) -> dict:
    """Extract the active Blender scene through the negotiated manifest path."""
    blender_scene = bpy.context.scene
    project_id = blender_scene.get("omb.project_id")
    if not isinstance(project_id, str):
        raise ValueError("scene is missing omb.project_id")

    scene_objects = list(blender_scene.objects)
    object_ids = {
        scene_object: _entity_id(scene_object, f"object {scene_object.name!r}")
        for scene_object in scene_objects
    }
    cameras = []
    lights = []
    animations = []
    for scene_object in scene_objects:
        object_id = object_ids[scene_object]
        if scene_object.type == "CAMERA":
            camera = scene_object.data
            cameras.append(
                {
                    "objectId": object_id,
                    "lens": float(camera.lens),
                    "sensorFit": _text(camera.sensor_fit),
                    "sensorWidth": float(camera.sensor_width),
                    "sensorHeight": float(camera.sensor_height),
                    "verticalFovRadians": float(camera.angle_y),
                    "clipStart": float(camera.clip_start),
                    "clipEnd": float(camera.clip_end),
                }
            )
            for target, animation_data in (
                ("object", scene_object.animation_data),
                ("cameraData", camera.animation_data),
            ):
                animation = _animation_snapshot(
                    object_id,
                    target,
                    animation_data,
                    identity_key="objectId",
                    include_handle_types=True,
                )
                if animation is not None:
                    animations.append(animation)
        elif scene_object.type == "LIGHT":
            light = scene_object.data
            light_entry = {
                "objectId": object_id,
                "lightType": _text(light.type),
                "color": _vector(light.color),
                "energy": float(light.energy),
                "spotSize": float(light.spot_size) if light.type == "SPOT" else None,
                "spotBlend": float(light.spot_blend) if light.type == "SPOT" else None,
            }
            if schema_version == 3:
                light_entry["areaSize"] = (
                    float(light.size) if light.type == "AREA" else None
                )
            lights.append(light_entry)

    fps_numerator, fps_denominator = rational_fps(
        int(blender_scene.render.fps), float(blender_scene.render.fps_base)
    )
    manifest_parts = dict(
        project_id=project_id,
        blender_version=_text(bpy.app.version_string),
        scene={
            "name": _text(blender_scene.name),
            "frameStart": int(blender_scene.frame_start),
            "frameEnd": int(blender_scene.frame_end),
            "fpsNumerator": fps_numerator,
            "fpsDenominator": fps_denominator,
            "activeCameraId": object_ids.get(blender_scene.camera),
        },
        render={
            "resolutionX": int(blender_scene.render.resolution_x),
            "resolutionY": int(blender_scene.render.resolution_y),
            "resolutionPercentage": int(blender_scene.render.resolution_percentage),
        },
        objects=[_manifest_object(scene_object) for scene_object in scene_objects],
        bones=_manifest_bones(scene_objects),
        cameras=cameras,
        lights=lights,
        markers=[
            {
                "name": _text(marker.name),
                "frame": int(marker.frame),
                "cameraId": object_ids.get(marker.camera),
            }
            for marker in blender_scene.timeline_markers
        ],
        selected_entity_ids=[
            object_ids[scene_object]
            for scene_object in scene_objects
            if scene_object.select_get()
        ],
        camera_animations=animations,
    )
    if schema_version == 3:
        stage_primitives, stage_materials = _stage_manifest_entries(
            scene_objects, object_ids
        )
        manifest = build_scene_manifest_v3(
            **manifest_parts,
            stage_primitives=stage_primitives,
            stage_materials=stage_materials,
        )
    else:
        manifest = build_scene_manifest(**manifest_parts)
    return finalize_scene_manifest(manifest)


def extract_scene_manifest_v2() -> dict:
    """Extract the active Blender scene through SceneManifestV2."""
    return _extract_scene_manifest(2)


def extract_scene_manifest_v3() -> dict:
    """Extract stage_scene state through the additive SceneManifestV3."""
    return _extract_scene_manifest(3)

def extract_scene_snapshot() -> dict:
    """Extract the active Blender scene into a Scene Snapshot v2 dictionary."""
    blender_scene = bpy.context.scene
    if blender_scene.render.fps_base != 1.0:
        raise UNSUPPORTED_FPS_BASE(
            f"fps_base must be 1.0, got {blender_scene.render.fps_base!r}"
        )

    scene_objects = list(blender_scene.objects)
    objects = [_object_snapshot(scene_object) for scene_object in scene_objects]
    normalized_names = [item["name"] for item in objects]
    if len(normalized_names) != len(set(normalized_names)):
        raise ValueError("object names must be unique after NFC normalization")
    cameras = [
        _camera_snapshot(scene_object)
        for scene_object in scene_objects
        if scene_object.type == "CAMERA"
    ]
    animations = []
    for scene_object in scene_objects:
        object_animation = _animation_snapshot(
            _text(scene_object.name), "object", scene_object.animation_data
        )
        if object_animation is not None:
            animations.append(object_animation)
        if scene_object.type == "CAMERA":
            camera_animation = _animation_snapshot(
                _text(scene_object.name), "cameraData", scene_object.data.animation_data
            )
            if camera_animation is not None:
                animations.append(camera_animation)

    return assemble_snapshot(
        scene={
            "name": _text(blender_scene.name),
            "frameStart": int(blender_scene.frame_start),
            "frameEnd": int(blender_scene.frame_end),
            "fps": int(blender_scene.render.fps),
            "activeCamera": _text(blender_scene.camera.name) if blender_scene.camera else None,
        },
        render={
            "resolutionX": int(blender_scene.render.resolution_x),
            "resolutionY": int(blender_scene.render.resolution_y),
            "resolutionPercentage": int(blender_scene.render.resolution_percentage),
        },
        objects=objects,
        cameras=cameras,
        markers=[
            {
                "name": _text(marker.name),
                "frame": int(marker.frame),
                "camera": _text(marker.camera.name) if marker.camera else None,
            }
            for marker in blender_scene.timeline_markers
        ],
        animations=animations,
    )


def write_scene_snapshot(output_path: Path) -> str:
    """Write a human-readable snapshot and print and return its revision."""
    snapshot = extract_scene_snapshot()
    output_path.write_text(
        json.dumps(snapshot, indent=2, ensure_ascii=False, allow_nan=False) + "\n",
        encoding="utf-8",
    )
    revision = snapshot_revision(snapshot)
    print(revision)
    return revision


def write_scene_manifest_v2(output_path: Path) -> str:
    """Write a real Blender-extracted SceneManifestV2 and return its revision."""
    scene_manifest = extract_scene_manifest_v2()
    output_path.write_text(
        json.dumps(scene_manifest, indent=2, ensure_ascii=False, allow_nan=False) + "\n",
        encoding="utf-8",
    )
    revision = scene_manifest["revisionId"]
    print(revision)
    return revision

File: packages/blender-director/src/manifest.ts
import { type Static, Type } from "typebox";
import { Parse } from "typebox/value";
import { canonicalRevision } from "./canonical.ts";

const NameSchema = Type.String({ minLength: 1, maxLength: 256 });
const NullableNameSchema = Type.Union([NameSchema, Type.Null()]);
const Vector2Schema = Type.Tuple([Type.Number(), Type.Number()]);
const Vector3Schema = Type.Tuple([Type.Number(), Type.Number(), Type.Number()]);
const QuaternionSchema = Type.Tuple([
	Type.Number(),
	Type.Number(),
	Type.Number(),
	Type.Number(),
]);

const SceneSchema = Type.Object(
	{
		name: NameSchema,
		frameStart: Type.Integer({ minimum: 0, maximum: 1_048_574 }),
		frameEnd: Type.Integer({ minimum: 0, maximum: 1_048_574 }),
		fps: Type.Integer({ minimum: 1, maximum: 240 }),
		activeCamera: NullableNameSchema,
	},
	{ additionalProperties: false },
);

const RenderSchema = Type.Object(
	{
		resolutionX: Type.Integer({ minimum: 1, maximum: 65_536 }),
		resolutionY: Type.Integer({ minimum: 1, maximum: 65_536 }),
		resolutionPercentage: Type.Integer({ minimum: 1, maximum: 100 }),
	},
	{ additionalProperties: false },
);

const SceneObjectSchema = Type.Object(
	{
		name: NameSchema,
		type: Type.String({ minLength: 1 }),
		parent: NullableNameSchema,
		visible: Type.Boolean(),
		location: Vector3Schema,
		rotationMode: Type.String({ minLength: 1 }),
		rotationQuaternion: QuaternionSchema,
		scale: Vector3Schema,
	},
	{ additionalProperties: false },
);

const CameraSchema = Type.Object(
	{
		name: NameSchema,
		lens: Type.Number({ exclusiveMinimum: 0 }),
		sensorFit: Type.Union([
			Type.Literal("AUTO"),
			Type.Literal("HORIZONTAL"),
			Type.Literal("VERTICAL"),
		]),
		sensorWidth: Type.Number({ exclusiveMinimum: 0 }),
		sensorHeight: Type.Number({ exclusiveMinimum: 0 }),
		verticalFovRadians: Type.Number({ exclusiveMinimum: 0, exclusiveMaximum: Math.PI }),
		clipStart: Type.Number({ exclusiveMinimum: 0 }),
		clipEnd: Type.Number({ exclusiveMinimum: 0 }),
	},
	{ additionalProperties: false },
);

const MarkerSchema = Type.Object(
	{
		name: NameSchema,
		frame: Type.Integer(),
		camera: NullableNameSchema,
	},
	{ additionalProperties: false },
);

const KeyframeSchema = Type.Object(
	{
		frame: Type.Number(),
		value: Type.Number(),
		interpolation: Type.String({ minLength: 1 }),
		handleLeft: Vector2Schema,
		handleRight: Vector2Schema,
	},
	{ additionalProperties: false },
);

const FCurveSchema = Type.Object(
	{
		dataPath: Type.String(),
		arrayIndex: Type.Integer({ minimum: 0 }),
		keyframes: Type.Array(KeyframeSchema),
	},
	{ additionalProperties: false },
);

const AnimationSchema = Type.Object(
	{
		objectName: NameSchema,
		target: Type.Union([Type.Literal("object"), Type.Literal("cameraData")]),
		fcurves: Type.Array(FCurveSchema),
	},
	{ additionalProperties: false },
);

export const SceneSnapshotSchema = Type.Object(
	{
		schemaVersion: Type.Literal(2),
		scene: SceneSchema,
		render: RenderSchema,
		objects: Type.Array(SceneObjectSchema),
		cameras: Type.Array(CameraSchema),
		markers: Type.Array(MarkerSchema),
		animations: Type.Array(AnimationSchema),
	},
	{ additionalProperties: false },
);

export type SceneSnapshot = Static<typeof SceneSnapshotSchema>;

export interface ProjectManifest {
	readonly revision: string;
	readonly snapshot: SceneSnapshot;
}

function compareCodePoints(left: string, right: string): number {
	const leftPoints = Array.from(left, (value) => value.codePointAt(0)!);
	const rightPoints = Array.from(right, (value) => value.codePointAt(0)!);
	for (let index = 0; index < Math.min(leftPoints.length, rightPoints.length); index += 1) {
		if (leftPoints[index] !== rightPoints[index]) return leftPoints[index]! - rightPoints[index]!;
	}
	return leftPoints.length - rightPoints.length;
}

function assertSorted<T>(values: readonly T[], compare: (left: T, right: T) => number, label: string): void {
	for (let index = 1; index < values.length; index += 1) {
		if (compare(values[index - 1]!, values[index]!) > 0) {
			throw new Error(`${label} must be in semantic order`);
		}
	}
}

function compareNullableNames(left: string | null, right: string | null): number {
	if (left === right) return 0;
	if (left === null) return -1;
	if (right === null) return 1;
	return compareCodePoints(left, right);
}

function validateSnapshot(snapshot: SceneSnapshot): void {
	if (snapshot.scene.frameStart > snapshot.scene.frameEnd) {
		throw new Error("scene.frameStart must not exceed scene.frameEnd");
	}

	const objects = new Map<string, (typeof snapshot.objects)[number]>();
	for (const object of snapshot.objects) {
		if (objects.has(object.name)) throw new Error(`duplicate object name: ${object.name}`);
		objects.set(object.name, object);
	}
	for (const object of snapshot.objects) {
		if (object.parent !== null && !objects.has(object.parent)) {
			throw new Error(`object ${object.name} has unknown parent: ${object.parent}`);
		}
	}

	const cameraNames = new Set<string>();
	for (const camera of snapshot.cameras) {
		if (cameraNames.has(camera.name)) throw new Error(`duplicate camera entry: ${camera.name}`);
		cameraNames.add(camera.name);
		if (objects.get(camera.name)?.type !== "CAMERA") {
			throw new Error(`camera ${camera.name} must reference a CAMERA object`);
		}
		if (camera.clipEnd <= camera.clipStart) {
			throw new Error(`camera ${camera.name} clipEnd must be greater than clipStart`);
		}
	}
	for (const object of snapshot.objects) {
		if (object.type === "CAMERA" && !cameraNames.has(object.name)) {
			throw new Error(`CAMERA object ${object.name} must have exactly one camera entry`);
		}
	}

	if (snapshot.scene.activeCamera !== null && !cameraNames.has(snapshot.scene.activeCamera)) {
		throw new Error(`scene.activeCamera references unknown camera: ${snapshot.scene.activeCamera}`);
	}
	for (const marker of snapshot.markers) {
		if (marker.camera !== null && !cameraNames.has(marker.camera)) {
			throw new Error(`marker ${marker.name} references unknown camera: ${marker.camera}`);
		}
	}

	for (const animation of snapshot.animations) {
		const object = objects.get(animation.objectName);
		if (object === undefined) {
			throw new Error(`animation references unknown object: ${animation.objectName}`);
		}
		if (animation.target === "cameraData" && object.type !== "CAMERA") {
			throw new Error(`cameraData animation requires a CAMERA object: ${animation.objectName}`);
		}
		assertSorted(
			animation.fcurves,
			(left, right) => compareCodePoints(left.dataPath, right.dataPath) || left.arrayIndex - right.arrayIndex,
			`animation ${animation.objectName} fcurves`,
		);
		for (const fcurve of animation.fcurves) {
			for (let index = 1; index < fcurve.keyframes.length; index += 1) {
				if (fcurve.keyframes[index - 1]!.frame >= fcurve.keyframes[index]!.frame) {
					throw new Error(`keyframe frames must be strictly increasing: ${animation.objectName}/${fcurve.dataPath}`);
				}
			}
		}
	}

	assertSorted(snapshot.objects, (left, right) => compareCodePoints(left.name, right.name), "objects");
	assertSorted(snapshot.cameras, (left, right) => compareCodePoints(left.name, right.name), "cameras");
	assertSorted(
		snapshot.markers,
		(left, right) =>
			compareCodePoints(left.name, right.name) ||
			left.frame - right.frame ||
			compareNullableNames(left.camera, right.camera),
		"markers",
	);
	assertSorted(
		snapshot.animations,
		(left, right) =>
			compareCodePoints(left.objectName, right.objectName) || compareCodePoints(left.target, right.target),
		"animations",
	);
}

export function parseSceneSnapshot(input: unknown): SceneSnapshot {
	const snapshot = Parse(SceneSnapshotSchema, input);
	validateSnapshot(snapshot);
	return snapshot;
}

export function createProjectManifest(snapshot: SceneSnapshot): ProjectManifest {
	return { revision: canonicalRevision(snapshot), snapshot };
}
